File: src/main/java/org/tindalos/principle/domain/detector/acd/ACDResult.java
package org.tindalos.principle.domain.detector.acd;

import org.tindalos.principle.domain.coredetector.CheckResult;
import org.tindalos.principle.domain.expectations.PackageCoupling;
import org.tindalos.principle.domain.expectations.cumulativedependency.DoubleThresholder;

public class ACDResult implements CheckResult {

    private final Double acd;
    private final Double rAcd;
    private final Double nCcd;
	private final PackageCoupling packageCouplingExpectation;

    public ACDResult(int cumulatedComponentDependency, int numOfComponents, PackageCoupling packageCoupling) {
        this.packageCouplingExpectation = packageCoupling;
		this.acd = (double) cumulatedComponentDependency / (double) numOfComponents;
        this.rAcd = this.acd / (double) numOfComponents;
        this.nCcd = this.acd / (double) numOfComponents;
    }

    public Double getACD() {
		return acd;
	}
    
    public Double getRACD() {
		return rAcd;
	}
    
    public Double getNCCD() {
		return nCcd;
	}
    
	public boolean expectationsFailed() {
		return greaterIfExists(acd, packageCouplingExpectation.getACD())
				|| greaterIfExists(rAcd, packageCouplingExpectation.getRACD())
				|| greaterIfExists(nCcd, packageCouplingExpectation.getNCCD());
	}

	private boolean greaterIfExists(Double actual, DoubleThresholder expectation) {
		if (expectation == null || expectation.getThreshold() == null) {
			return false;
		}
		return actual > expectation.getThreshold();
	}



}

File: src/main/java/org/tindalos/principle/domain/detector/acd/ACDResult.scala
package org.tindalos.principle.domain.detector.acd

import org.tindalos.principle.domain.expectations.PackageCoupling
import org.tindalos.principle.domain.coredetector.AnalysisResult
import org.tindalos.principle.domain.expectations.cumulativedependency.DoubleThresholder

class ACDResult(val cumulatedComponentDependency: Int, val numOfComponents: Int, val packageCoupling: PackageCoupling) extends AnalysisResult {

  val acd = cumulatedComponentDependency.toDouble / numOfComponents.toDouble
  val rAcd = acd / numOfComponents.toDouble
  val nCcd = acd / numOfComponents.toDouble

  override def expectationsFailed() =
    greaterIfExists(rAcd, packageCoupling.racd) ||
    greaterIfExists(nCcd, packageCoupling.nccd)

  private def greaterIfExists(actual: Double, expectation: DoubleThresholder) =
    if (expectation == null || expectation.threshold == Double.NaN) false
    else actual > expectation.threshold

  def getRACDThreshold() =
    if (packageCoupling.racd == null) 999d
    else packageCoupling.racd.threshold
}